src/inference/graph_inference.h

```diff
@@ -113,7 +119,7 @@ struct FactorFeaturesLevel {
   }
 
   std::vector<std::pair<double, Factor>> factor_features;
-  std::unordered_map<int, std::shared_ptr<FactorFeaturesLevel>> next_level;
+  google::dense_hash_map<int, std::shared_ptr<FactorFeaturesLevel>> next_level;
```

Comment:
I highly doubt dense_hash_map is faster than unordered_map (or even map) here. And you may have bugs if a label is -1 (the empty key).